quickwit/quickwit-serve/src/decompression.rs

```diff
@@ -78,10 +79,29 @@ pub(crate) struct UnsupportedEncoding(String);
 impl Reject for UnsupportedEncoding {}
 
 /// Custom filter for optional decompression
-pub(crate) fn get_body_bytes() -> impl Filter<Extract = (Bytes,), Error = warp::Rejection> + Clone {
+pub(crate) fn get_body_bytes() -> impl Filter<Extract = (Body,), Error = warp::Rejection> + Clone {
     warp::header::optional("content-encoding")
         .and(warp::body::bytes())
         .and_then(|encoding: Option<String>, body: Bytes| async move {
-            decompress_body(encoding, body).await
+            decompress_body(encoding, body).await.map(Body::from)
         })
 }
+
+pub(crate) struct Body {
```

Comment:
`Deref` (in moderation) can be useful for that kind of job.
By implementing `Deref<Target=Bytes>` on a `struct TrackedBytes { bytes: Bytes, guard }`  you can both get the tracking and all of the method of bytes.

I only use it on small object like this because the downside is that it is a bit magic.